vparser/__init__.py

```diff
@@ -0,0 +1,166 @@
+'''
+module : vparser
+
+Query API経由でJSONを得て、ライブ動画IDを取り出す。
+'''
+
+import json
+import requests
+from httpreq import HttpRequest
+
+'''
+JSONパース用のpath
+'''
+p_tabroot = [
+    'response',
+    'contents',
+    'singleColumnBrowseResultsRenderer',
+    'tabs'
+]
+
+sx_tabtitle = [
+    'tabRenderer',
+    'title'
+]
+
+sx_contents_part = [
+    'tabRenderer',
+    'content',
+    'sectionListRenderer',
+    'contents',
+    0,
+    'itemSectionRenderer',
+    'contents'
+]
+
+p_videotab = [
+    'response',
+    'contents',
+    'singleColumnBrowseResultsRenderer',
+    'tabs',
+    1,
+    'tabRenderer',
+    'title'
+]
+
+p_contents = [
+    'response',
+    'contents',
+    'singleColumnBrowseResultsRenderer',
+    'tabs',
+    1,
+    'tabRenderer',
+    'content',
+    'sectionListRenderer',
+    'contents',
+    0,
+    'itemSectionRenderer',
+    'contents'
+]
+
+px_vid = [
+    'compactVideoRenderer',
+    'videoId',
+]
+
+px_title = [
+    'compactVideoRenderer',
+    'title',
+    'runs',
+    0,
+    'text'
+]
+
+px_status = [
+    'compactVideoRenderer',
+    'thumbnailOverlays',
+    0,
+    'thumbnailOverlayTimeStatusRenderer',
+    'style'
+]
+
+'''
+404エラーのJSON path
+'''
+x_404_error = [
+    'response',
+    'responseContext',
+    'errors',
+    0
+]
+
+
+class InvalidChannelIDException(requests.exceptions.RequestException):
+    '''
+    指定したチャンネルIDのページが不存在の場合に投げる例外
+    '''
+    pass
+
+
+def get_source_json(req:HttpRequest, channel_id):
+    '''
+    Query APIを叩いてライブ動画IDのリストを含むJSONデータを取り出す.
+
+    戻り値:(str):動画IDのリストを含むJSON文字列。
+        リクエスト時にネットワークエラー等のエラーが発生した場合は
+        空の文字列 （''）を返す。
+    '''
+    url = f'https://m.youtube.com/channel/{channel_id}/videos'
+    params = {'view':2, 'flow':'list', 'pbj':1}
+    response = req.get(url=url, params=params)
+    return response.text
+
+def extract_video_ids(source_json:str):
+    '''
+    JSONからライブ動画のIDのリストを抽出する。
+    
+    戻り値:(List[str]):ライブ動画のIDのリスト。
+        ライブ動画なし,またはJSONが不正なフォーマット/JSON取得時にエラー発生の場合
+        空のリスト([])を返す。
+    '''
+    try:
+        source_dic = json.loads(source_json)
+    except json.JSONDecodeError:
+        print('JSONのパースに失敗しました')
+        return []
+    # チャンネルページが存在しないエラーを示すJSONを
+    # 受け取った場合は、例外を呼び出し側(GUI)に伝播させる。
+    if _getitem(source_dic, x_404_error) == (
+        'This channel does not exist.'
+    ):
+        raise InvalidChannelIDException()
+    return _get_live_vids(source_dic)
+
+def _get_live_vids(dic):
+    contents = _getitem(dic, p_contents) or _find_videos_tab(dic) or []
```

Comment:
I think `dic` passed into _getitem() will be modified internally and be different from the `dic` received in _find_video_tab().

Reply:
Oh, I'm wrong. I'll delete this comment. Sorry for my bothering.

Reply:
Penetrating insight!

I created test cases in my forked repository.
https://github.com/taizan-hokuto/mikochiku_alarm/tree/feature/test_case

To execute test, input in ter below commands:
```terminal
pip install -r test_requirements.txt

pytest test
```
In `/test/testadata/` directory, `test.json` and `test_cracked.json` are just diferrent that the index position of 'Videos' tab.

see:

https://github.com/taizan-hokuto/mikochiku_alarm/commit/543c2fcbc0a300a5d842afa2b2436f7c4fbc51e7


If the contents of `dic` passed into `_get_live_vids` is modified by the first function `_getitem(dic, p_contents)`, then second function `_find_videos_tab(dic)` would return None, and `_get_live_vids` would return empty list ([]), and the second test case would fail.

But the tests all success.

The following is my guess:
When a function is called, the pointer for argument object of caller is newly copied at callee function (Of course, the address pointed to is the same in both pointers) .
Changing the address of the callee pointer does not affect the caller pointer itself.

The `dic` (we call "dic_a") in `called _getitem` function is newly allocated when called. 
The target of dic_a changes changes as it follow the items described in the list.

On the other hand, `dic`(we call "dic_b") in `_get_live_vids` always points the source_dic.

Reply:
So sorry...That's very kind of you to take so much time to explain the reasons.
I apologize for my mistake!